kf/bpf.go

```diff
@@ -1098,6 +1106,123 @@ func (b *BPF) VerifyMetricsMapsVanish() error {
 	return err
 }
 
+// LoadXDPAttachProgram - Load and attach xdp root program or any xdp program when chaining is disabled
+func (b *BPF) LoadXDPAttachProgram(ifaceName string, eBPFProgram *BPF) error {
+	iface, err := net.InterfaceByName(ifaceName)
+	if err != nil {
+		log.Fatal().Msgf("look up network iface %q: %s", ifaceName, err)
+	}
+
+	CollectionRef, err := ebpf.LoadCollection(eBPFProgram.Program.ObjectFile)
+	if err != nil {
+		return fmt.Errorf("%s: loading of xdp root failed", eBPFProgram.Program.ObjectFile)
+	}
+
+	b.ProgMapCollection = CollectionRef
+	bpfRootProg := CollectionRef.Programs[eBPFProgram.Program.EntryFunctionName]
+	bpfRootMap := CollectionRef.Maps[eBPFProgram.Program.MapName]
+
+	// Pinning root map
+	if b.hostConfig.BpfChainingEnabled {
+		rootArrayMapFileName := filepath.Join(b.hostConfig.BpfMapDefaultPath, eBPFProgram.Program.MapName)
+		if err := bpfRootMap.Pin(rootArrayMapFileName); err != nil {
+			return fmt.Errorf("%s:failed to pin the map", rootArrayMapFileName)
+		}
+	}
+
+	b.XDPLink, err = link.AttachXDP(link.XDPOptions{
+		Program:   bpfRootProg,
+		Interface: iface.Index,
+	})
+
+	if err != nil {
+		return fmt.Errorf("could not attach xdp program %s to interface %s : %v", b.Program.Name, ifaceName, err)
+	}
+
+	progInfo, err := bpfRootProg.Info()
+	if err != nil {
+		return fmt.Errorf("could not get program info of %s to interface %s : %v", b.Program.Name, ifaceName, err)
+	}
+
+	ok := false
+	b.ProgID, ok = progInfo.ID()
+	if !ok {
+		return fmt.Errorf("failed to fetch the xdp program %s to interface %s : %v", b.Program.Name, ifaceName, err)
+	}
+
+	ebpfInfo, err := bpfRootMap.Info()
+	if err != nil {
+		return fmt.Errorf("fetching map info failed for xdp program %s to interface %s : %v", b.Program.Name, ifaceName, err)
+	}
+
+	b.ProgMapID, ok = ebpfInfo.ID()
+	if !ok {
+		return fmt.Errorf("fetching map id failed for xdp program %s to interface %s : %v", b.Program.Name, ifaceName, err)
+	}
+
+	return nil
+}
+
+// UnloadProgram - Unload or detach the program from the interface and close all the program resources
+// TODO: Before unloading make sure user program is stopped to avoid any errors
```

Comment:
Is there a github issue filed to track this TODO?

Reply:
This particular pull request is solely focused on loading root programs. However, in the next pull request, there will be a more structured flow for loading xdp and tc programs, starting with user programs and vice versa.

Reply:
Is there a github issue filed to track that work?_https://github.com/l3af-project/l3afd/blob/main/docs/CONTRIBUTING.md says "It is recommended prior to working on a PR to submit an issue in github for the change you want to make describing the change and context around it.  ..."